src/SparseMatrix.chpl

```diff
@@ -0,0 +1,346 @@
+    return A;
+  }
+
+  module SpsMatUtil {
+    // The following are routines that should arguably be supported directly
+    // by the LayoutCS and SparseBlockDist modules themselves
+    //
+    //  public use LayoutCSUtil, SparseBlockDistUtil;
+
+    use BlockDist, LayoutCS, Map, Random;
+
+    enum layout { CSR, CSC };
+    public use layout;
+
+    config const seed = 0,
+                printSeed = seed == 0;
+
+    var rands = if seed == 0 then new randomStream(real)
+                          else new randomStream(real, seed);
+
+    // print library-selected seed, for reproducibility
+    //
+    if printSeed then
+      writeln("Using seed: ", rands.seed);
+
+    record sparseMatDat {
+      forwarding var m: map(2*int, int);
+
+      proc ref add(idx: 2*int, val: int) {
+        if val != 0 {
+          if m.contains(idx) {
+            m[idx] += val;
+          } else {
+            m.add(idx, val);
+          }
+        }
+      }
+    }
+
+
+    // create a local random sparse matrix within the space of 'Dom' of
+    // the given density and layout.  If distributed is true, this will
+    // be a block-distributed sparse matrix, otherwise it'll be local.
+    //
+    proc randSparseDomain(parentDom, density, param layout, param distributed)
+    where distributed == false {
+
+      var SD: sparse subdomain(parentDom) dmapped new dmap(new CS(compressRows=(layout==CSR)));
+
+      for (i,j) in parentDom do
+        if rands.next() <= density then
+          SD += (i,j);
+
+      return SD;
+    }
+
+    proc randSparseDomain(parentDom, density, param layout, param distributed)
+    where distributed == true {
+      const locsPerDim = sqrt(numLocales:real): int,
+            grid = {0..<locsPerDim, 0..<locsPerDim},
+            localeGrid = reshape(Locales[0..<grid.size], grid);
+
+
+      if grid.size != numLocales then
+        writeln("Warning: Only using ", grid.size, " of ", numLocales,
+                " locales");
+
+      // writeln(grid);
+
+      type layoutType = CS(compressRows=(layout==CSR));
+      const DenseBlkDom = parentDom dmapped new blockDist(boundingBox=parentDom,
+                                                    targetLocales=localeGrid,
+                                                    sparseLayoutType=layoutType);
+
+      var SD: sparse subdomain(DenseBlkDom);
+
+      for (i,j) in parentDom do
+        if rands.next() <= density then
+          SD += (i,j);
+
+      return SD;
+    }
+
+
+    proc emptySparseDomLike(Mat) {
+      var ResDom: sparse subdomain(Mat.domain.parentDom);
+      return ResDom;
+    }
+
+
+    // print out a sparse matrix (in a dense format)
+    //
+    proc writeSparseMatrix(msg, Arr) {
+      const ref SparseDom = Arr.domain,
+                DenseDom = SparseDom.parentDom;
+
+      writeln(msg);
+
+      for r in DenseDom.dim(0) {
+        for c in DenseDom.dim(1) {
+          write(Arr[r,c], " ");
+        }
+        writeln();
+      }
+      writeln();
+    }
```

Comment:
We could just have the Arkouda `__str__` function call this in a truncated format for the prints.

Reply:
I think we discussed that we should not print the array as if it were dense (Which is what this function does), rather we want to represent the arrays which have the non IRV values, and the corresponding indices (depending on if the matrix is CSC or CSR for the right representation there).